handler/src/test/java/io/netty/handler/flow/FlowControlHandlerTest.java

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version
+ * 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.handler.flow;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.hamcrest.collection.IsIterableContainingInOrder;
+import org.junit.Test;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.embedded.EmbeddedChannel;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class FlowControlHandlerTest {
+
+    /**
+     * This test is showing the default behavior. Auto reading is turned off but
+     * we still receive messages because ChannelHandler's are not subject to it.
+     */
+    @Test
+    public void testNettyAutoReadOff() {
+
+        final AtomicInteger counter = new AtomicInteger();
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg)
+                    throws Exception {
+                counter.incrementAndGet();
+            }
+        };
+
+        EmbeddedChannel channel = new EmbeddedChannel(handler);
```

Comment:
a reason why I suggest not using `EmbeddedChannel` is because `FlowControlHandler` is overriding methods (read) and overloading/re-using concepts (autoRead()) also used at a lower levels. It feels like a good idea to have the unit test live in the "real environment" as much as possible to ensure these layers play nicely together, instead of `EmbeddedChannel` which simulates these behaviors in a way that is not entirely consistent (it doesn't use autoRead).